File: java_src/org/tarot/billing/BillingManager.java
package org.tarot.billing;

import android.app.Activity;
import android.util.Log;

import com.android.billingclient.api.AcknowledgePurchaseParams;
import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.ConsumeParams;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.android.billingclient.api.QueryProductDetailsParams;
import com.android.billingclient.api.QueryPurchasesParams;
import com.android.billingclient.api.ProductDetails;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Google Play Billing v8 manager for Kivy (non-consumable products).
 * Python usage via pyjnius:
 * BillingManager.init(activity)
 * BillingManager.purchase(activity, "remove_ads")
 * BillingManager.hasPurchased("remove_ads")
 */
public final class BillingManager {
    private static final String TAG = "BillingManager";

    private static BillingClient billingClient = null;
    private static boolean billingReady = false;
    private static final Map<String, Purchase> purchasesMap = new HashMap<>();

    private BillingManager() {}

    public static void init(Activity activity) {
        try {
            if (billingClient != null && billingClient.isReady()) {
                Log.d(TAG, "init(): already ready");
                billingReady = true;
                queryPurchasesAsync();
                return;
            }
        billingClient = BillingClient.newBuilder(activity)
            .enablePendingPurchases(
                com.android.billingclient.api.PendingPurchasesParams
                    .newBuilder()
                    .enableOneTimeProducts()
                    .build()
            )
            .setListener(BillingManager::onPurchasesUpdated)
            .build();

            billingClient.startConnection(new BillingClientStateListener() {
                @Override
                public void onBillingSetupFinished(BillingResult billingResult) {
                    if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                        Log.d(TAG, "Billing setup finished OK");
                        billingReady = true;
                        queryPurchasesAsync();
                    } else {
                        Log.w(TAG, "Billing setup finished code=" + billingResult.getResponseCode() + " msg=" + billingResult.getDebugMessage());
                        billingReady = false;
                    }
                }

                @Override
                public void onBillingServiceDisconnected() {
                    Log.w(TAG, "Billing service disconnected");
                    billingReady = false;
                }
            });
        } catch (Throwable t) {
            Log.e(TAG, "init failed: " + t.getMessage(), t);
        }
    }

    public static boolean isBillingReady() {
        return billingReady;
    }

    public static void purchase(Activity activity, String productId) {
        if (!billingReady) {
            Log.w(TAG, "purchase() called but billing not ready");
            return;
        }
        try {
            QueryProductDetailsParams params = QueryProductDetailsParams.newBuilder()
                    .setProductList(
                            java.util.Collections.singletonList(
                                    QueryProductDetailsParams.Product.newBuilder()
                                            .setProductId(productId)
                                            .setProductType(BillingClient.ProductType.INAPP)
                                            .build()
                            )
                    )
                    .build();

            billingClient.queryProductDetailsAsync(params, (billingResult, productDetailsResult) -> {
                java.util.List<ProductDetails> productDetailsList = (productDetailsResult != null)
                        ? productDetailsResult.getProductDetailsList()
                        : null;
                if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK && productDetailsList != null && !productDetailsList.isEmpty()) {
                    ProductDetails pd = productDetailsList.get(0);
                    BillingFlowParams.ProductDetailsParams detailsParams = BillingFlowParams.ProductDetailsParams.newBuilder()
                            .setProductDetails(pd)
                            .build();
                    BillingFlowParams flowParams = BillingFlowParams.newBuilder()
                            .setProductDetailsParamsList(java.util.Collections.singletonList(detailsParams))
                            .build();
                    BillingResult launchRes = billingClient.launchBillingFlow(activity, flowParams);
                    Log.d(TAG, "launchBillingFlow response=" + launchRes.getResponseCode() + " msg=" + launchRes.getDebugMessage());
                } else {
                    Log.e(TAG, "queryProductDetailsAsync failed: " + billingResult.getDebugMessage());
                }
            });
        } catch (Throwable t) {
            Log.e(TAG, "purchase failed: " + t.getMessage(), t);
        }
    }

    public static void onPurchasesUpdated(BillingResult billingResult, List<Purchase> purchases) {
        int code = billingResult.getResponseCode();
        if (code == BillingClient.BillingResponseCode.OK) {
            if (purchases != null) handlePurchases(purchases);
        } else if (code == BillingClient.BillingResponseCode.USER_CANCELED) {
            Log.d(TAG, "User canceled purchase");
        } else {
            Log.e(TAG, "Purchase failed code=" + code + " msg=" + billingResult.getDebugMessage());
        }
    }

    private static void handlePurchases(List<Purchase> purchases) {
        try {
            for (Purchase p : purchases) {
                for (String pid : p.getProducts()) {
                    purchasesMap.put(pid, p);
                }
                if (p.getPurchaseState() == Purchase.PurchaseState.PURCHASED && !p.isAcknowledged()) {
                    AcknowledgePurchaseParams ack = AcknowledgePurchaseParams.newBuilder()
                            .setPurchaseToken(p.getPurchaseToken())
                            .build();
                    billingClient.acknowledgePurchase(ack, result ->
                            Log.d(TAG, "Acknowledge result code=" + result.getResponseCode() + " msg=" + result.getDebugMessage())
                    );
                }
            }
        } catch (Throwable t) {
            Log.e(TAG, "handlePurchases failed: " + t.getMessage(), t);
        }
    }

    private static void queryPurchasesAsync() {
        try {
            QueryPurchasesParams params = QueryPurchasesParams.newBuilder()
                    .setProductType(BillingClient.ProductType.INAPP)
                    .build();
            billingClient.queryPurchasesAsync(params, (billingResult, purchaseList) -> {
                if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                    purchasesMap.clear();
                    if (purchaseList != null) {
                        for (Purchase p : purchaseList) {
                            for (String pid : p.getProducts()) purchasesMap.put(pid, p);
                        }
                    }
                    Log.d(TAG, "queryPurchasesAsync loaded=" + purchasesMap.size());
                } else {
                    Log.w(TAG, "queryPurchasesAsync failed: " + billingResult.getDebugMessage());
                }
            });
        } catch (Throwable t) {
            Log.e(TAG, "queryPurchasesAsync failed: " + t.getMessage(), t);
        }
    }

    public static boolean hasPurchased(String productId) {
        Purchase p = purchasesMap.get(productId);
        return p != null && p.getPurchaseState() == Purchase.PurchaseState.PURCHASED;
    }

    public static void consume(String productId) {
        try {
            Purchase p = purchasesMap.get(productId);
            if (p == null) {
                Log.w(TAG, "consume: no purchase for " + productId);
                return;
            }
            ConsumeParams params = ConsumeParams.newBuilder().setPurchaseToken(p.getPurchaseToken()).build();
            billingClient.consumeAsync(params, (billingResult, outToken) -> {
                Log.d(TAG, "consumeAsync code=" + billingResult.getResponseCode() + " msg=" + billingResult.getDebugMessage());
                if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                    purchasesMap.remove(productId);
                }
            });
        } catch (Throwable t) {
            Log.e(TAG, "consume failed: " + t.getMessage(), t);
        }
    }
}

File: java_src/org/tarot/billing/BillingManager.kt
package org.tarot.billing

import android.app.Activity
import android.util.Log
import com.android.billingclient.api.*

object BillingManager : PurchasesUpdatedListener {
    private const val TAG = "BillingManager"

    private var billingClient: BillingClient? = null
    private var billingReady: Boolean = false
    private val purchasesMap: MutableMap<String, Purchase> = mutableMapOf()

    @JvmStatic
    fun init(activity: Activity) {
        try {
            if (billingClient != null && billingClient?.isReady == true) {
                Log.d(TAG, "init() called but billingClient already ready")
                billingReady = true
                queryPurchasesAsync()
                return
            }

            billingClient = BillingClient.newBuilder(activity)
                .enablePendingPurchases()
                .setListener(this)
                .build()

            billingClient?.startConnection(object : BillingClientStateListener {
                override fun onBillingSetupFinished(billingResult: BillingResult) {
                    if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                        Log.d(TAG, "Billing setup finished OK")
                        billingReady = true
                        queryPurchasesAsync()
                    } else {
                        Log.w(TAG, "Billing setup finished with code: ${billingResult.responseCode} message=${billingResult.debugMessage}")
                        billingReady = false
                    }
                }

                override fun onBillingServiceDisconnected() {
                    Log.w(TAG, "Billing service disconnected")
                    billingReady = false
                }
            })
        } catch (t: Throwable) {
            Log.e(TAG, "init failed: ${t.message}", t)
        }
    }

    @JvmStatic
    fun isBillingReady(): Boolean = billingReady

    @JvmStatic
    fun purchase(activity: Activity, productId: String) {
        if (!billingReady) {
            Log.w(TAG, "purchase() called but billing not ready")
            return
        }

        try {
            val product = BillingClient.Product.newBuilder()
                .setProductId(productId)
                .setProductType(BillingClient.ProductType.INAPP)
                .build()

            val params = QueryProductDetailsParams.newBuilder()
                .setProductList(listOf(product))
                .build()

            billingClient?.queryProductDetailsAsync(params) { billingResult, productDetailsList ->
                if (billingResult.responseCode == BillingClient.BillingResponseCode.OK && !productDetailsList.isNullOrEmpty()) {
                    val productDetails = productDetailsList[0]

                    // Prepare flow params
                    val productDetailsParams = BillingFlowParams.ProductDetailsParams.newBuilder()
                        .setProductDetails(productDetails)
                        .build()

                    val flowParams = BillingFlowParams.newBuilder()
                        .setProductDetailsParamsList(listOf(productDetailsParams))
                        .build()

                    billingClient?.launchBillingFlow(activity, flowParams)?.let { result ->
                        Log.d(TAG, "launchBillingFlow result: ${result.responseCode} ${result.debugMessage}")
                    }
                } else {
                    Log.e(TAG, "ProductDetails not found or error: ${billingResult.debugMessage}")
                }
            }
        } catch (t: Throwable) {
            Log.e(TAG, "purchase failed: ${t.message}", t)
        }
    }

    override fun onPurchasesUpdated(billingResult: BillingResult, purchases: MutableList<Purchase>?) {
        when (billingResult.responseCode) {
            BillingClient.BillingResponseCode.OK -> {
                if (purchases != null) {
                    handlePurchases(purchases)
                }
            }
            BillingClient.BillingResponseCode.USER_CANCELED -> {
                Log.d(TAG, "User cancelled purchase")
            }
            else -> {
                Log.e(TAG, "Purchase failed: ${billingResult.debugMessage}")
            }
        }
    }

    private fun handlePurchases(purchases: List<Purchase>) {
        try {
            for (purchase in purchases) {
                // store
                for (productId in purchase.products) {
                    purchasesMap[productId] = purchase
                }

                if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED && !purchase.isAcknowledged) {
                    val ackParams = AcknowledgePurchaseParams.newBuilder()
                        .setPurchaseToken(purchase.purchaseToken)
                        .build()
                    billingClient?.acknowledgePurchase(ackParams) { ackResult ->
                        Log.d(TAG, "Acknowledge result: ${ackResult.responseCode} ${ackResult.debugMessage}")
                    }
                }
            }
        } catch (t: Throwable) {
            Log.e(TAG, "handlePurchases failed: ${t.message}", t)
        }
    }

    private fun queryPurchasesAsync() {
        try {
            val params = QueryPurchasesParams.newBuilder()
                .setProductType(BillingClient.ProductType.INAPP)
                .build()

            billingClient?.queryPurchasesAsync(params) { billingResult, purchaseList ->
                if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                    purchasesMap.clear()
                    purchaseList?.forEach { p ->
                        for (productId in p.products) purchasesMap[productId] = p
                    }
                    Log.d(TAG, "queryPurchasesAsync: loaded ${purchasesMap.size} purchases")
                } else {
                    Log.w(TAG, "queryPurchasesAsync failed: ${billingResult.debugMessage}")
                }
            }
        } catch (t: Throwable) {
            Log.e(TAG, "queryPurchasesAsync failed: ${t.message}", t)
        }
    }

    @JvmStatic
    fun hasPurchased(productId: String): Boolean {
        val p = purchasesMap[productId] ?: return false
        return p.purchaseState == Purchase.PurchaseState.PURCHASED
    }

    @JvmStatic
    fun consume(productId: String) {
        try {
            val purchase = purchasesMap[productId] ?: run {
                Log.w(TAG, "consume: no purchase found for $productId")
                return
            }
            val token = purchase.purchaseToken
            val consumeParams = ConsumeParams.newBuilder().setPurchaseToken(token).build()
            billingClient?.consumeAsync(consumeParams) { billingResult, outToken ->
                Log.d(TAG, "consumeAsync: ${billingResult.responseCode} ${billingResult.debugMessage}")
                if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                    purchasesMap.remove(productId)
                }
            }
        } catch (t: Throwable) {
            Log.e(TAG, "consume failed: ${t.message}", t)
        }
    }
}
